### DISCUSSION TITLE
Dealing with hidden staves

### DISCUSSION THREAD

**Discussion by jonarnoldmusic** [Audiveris/audiveris#546] [Q&A] [answered]
Thanks for all the work on Audiveris! I love having an open source means of OMR. 

I work mostly with choral music and often have a piece that starts with a piano introduction in which the voices are hidden in the first system. It also may happen in the middle of the piece- a system with piano only and no voices. When this happens, the measures of the piano and the voices are not aligned in export. How can I fix this?

**Comment by hbitteur** [comment]
The voices can come and go from one measure to the other. 

At export time, Audiveris can create virtual staves on-the-fly, made of a sequence of measures filled with whole rests, precisely to keep voices alignment in the exported data.

If not, there must be a bug somewhere.
Please provide your .omr file (or your input image file) so that we can check what happened in this case.

**Reply by jonarnoldmusic** [reply]
Thanks for your help. I've attached the PDF and OMR in a zip file. The issue with staves is on page 2 and 7. 

Also, I chose to export xml from each page individually because on page 4, the bracket for the voices in the second system was not identified correctly. I'm not sure how to fix that.
[systemError.zip](https://github.com/Audiveris/audiveris/files/7877567/systemError.zip)

**Comment by hbitteur** [comment, marked as answer]
The curly bracket in the bottom left of page 4 was not recognized. And since the role of such brace is to indicate that the embraced staves belong to the same part, Audiveris engine thought the 2 staves were separate parts. And since logical parts from one system to the next need to exhibit the same number of contained staves, the engine thought these two 1-staff parts could not belong to the same logical 2-staff part found at the end of system above. This explains your problems when exporting the results to MusicXML, which is done ... part by part!
Small problem, big consequence!

To fix this, simply point to the underlying glyph and in the shape board (found in right side column), select the "Barlines" palette. In this palette, you'll find the BRACE shape button. A double-click on this button, and your brace glyph gets assigned the BRACE shape. You'll be prompted about the target staff for this brace, select staff above.
Done!

Now, when exporting the whole score (you no longer need to export page by page), I got an exception on every page. Diving in the code, I discovered this was due to looking for potential beams related to a rest chord in the dummy measures inserted for voices alignment.
To make a long story short, since 5.2.4 a rest chord can be interleaved with other head chords in a beam. This is not frequent but had to be handled. Unfortunately, this beam check raises an exception when performed on a dummy chord!

I just fixed the bug in the development branch. You have two possibilities:
- Pull the development branch and build Audiveris on your own. See the handbook on how to build from sources.
- Or wait for a new release that contains the bug fix. I noticed you are using 5.2.4 on Windows. Since other people might encounter that same bug I introduced in November, I will try to publish a fix as 5.2.5 in the coming days.

**Reply by jonarnoldmusic** [reply]
Thanks for your help. It sounds like the bug you addressed will fix the first issue I raised about the hidden staves. I've taken care of my score, so I don't actually need the fix right now, but I'm glad it will be available as I have this issue with a lot of scores.

Thanks for your help with page 4. When I add the curly brace to the piano staves, there's a dialog that says "Select staff above or below." In this case with a brace in the margin, it's not clear what that's asking. Is it asking what to align the new part with or is it asking what staves to put inside the brace? It seems like the former as clicking the up arrow is what got me the desired solution. I think this dialog box could be more clear. 

Also, how is it that the curly brace on staves 11 and 12 affects the grouping of staff 7? Without the brace, staff 7 is not aligned with the other vocal parts in that system, but adding the brace groups them properly even though the system bracket is still not properly detected.

**Reply by jonarnoldmusic** [reply]
It seems that manually adding the bracket to staves 7-10 doesn't actually fix the problem- staff 7 still appears at the beginning of the exported excerpt. That doesn't make sense.

**Comment by hbitteur** [comment]
Regarding the popup dialog about which staff to select, I agree this is somewhat confusing for the case of a side brace.
This dialog is meant for the cases where the engine is not able to infer the related staff for an inserted symbol, hence it simply asks the user! The answer is generally rather obvious for a human being.
For the case at hand (a brace that embraces 2 staves, as it is often the case for a brace by the way), the engine was not smart enough to detect what to do with this symbol located right in the middle of the 2 staves! I need to fix that, and let the engine select the upper staff when dealing with a brace.

**Comment by hbitteur** [comment]
> Also, how is it that the curly brace on staves 11 and 12 affects the grouping of staff 7? Without the brace, staff 7 is not aligned with the other vocal parts in that system, but adding the brace groups them properly even though the system bracket is still not properly detected.

Raw engine output for this page 4 is a sequence of 12 staves gathered in 2 systems:
System 1:
- Staff 1
- Staff 2
- Staff 3
- Staff 4
- Staff 5, linked to staff 6 below
- Staff 6, linked to staff 5 above

System 2:
- Staff 7
- Staff 8
- Staff 9
- Staff 10
- Staff 11 (*Not linked with staff 12 below!*)
- Staff 12 (*Not linked with staff 11 above!*)

A staff not linked to another staff corresponds to an isolated part (a single instrument).
The engine tries to relate the parts in system 1 with the parts in system 2.
So parts for staves 1,2,3,4 relate to parts for staves 7,8,9,10 respectively. Fine.
But when the engine processes the fifth part at end of system 1 (made of staves 5 & 6) it can't find a suitable corresponding part at end of system 2, which exhibits 2 parts, each with its own staff (11 for upper part, 12 for lower part)

These part numbers, called "logical parts", are processed in the scope of the whole score (I should say movement to be more precise, but here we have a single movement in the score). For example you can find this fifth part in every system since the beginning of the score, until we hit the 2nd system of page 4. Suddenly, the engine is obliged to create two new parts (for stave 11 and for staff 12) because these "parts" have no corresponding parts early in the score. These new parts are named part 6 and part 7 respectively.

And since the export to MusicXML format is done "partwise", you get a separate XML fragment for every part. "Part" 6 gives a separate fragment. "Part" 7 gives another separate fragment.

Manually linking the staves 11 and 12 with a brace allows the engine to stay with its original 5 parts all score long.

**Comment by hbitteur** [comment]
@ace214 
I just pushed the 5.2.5 release which fixes these problems (bug on dummy-rests and spurious dialog on staff target for a brace)
Please download and install this new release (or pull and build by yourself)

**Comment by hbitteur** [comment]
I didn't notice, but we should have exchanged these messages on the **Issues** forum rather than on this **Discussions** forum.
This would have made them visible for the other users on the Issues forum.
Too bad...
